node/src/Pos/Wallet/Web/Pending/Functions.hs

```diff
@@ -0,0 +1,174 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+
+module Pos.Wallet.Web.Pending.Functions
+    ( reevaluateUncertainPtxs
+    , mkHashMap
+    ) where
+
+import           Universum
+
+import           Control.Lens                 (has, to, folded, _Right)
+import qualified Data.HashMap.Strict          as HM
+import qualified Data.HashSet                 as HS
+import           Formatting (sformat, (%))
+import           Serokell.Util (listJson)
+import           System.Wlog (logInfo, WithLogger)
+
+import           Pos.Block.Core               (Block, mainBlockTxPayload)
+import           Pos.Crypto                   (WithHash (..), hash)
+import           Pos.Core.Configuration       (HasConfiguration, genesisHash)
+import           Pos.Core.Types               (ChainDifficulty)
+import           Pos.Core.Class               (difficultyL, headerHash, prevBlockL)
+import qualified Pos.DB.DB                    as DB
+import qualified Pos.DB.Block                 as DB
+import           Pos.Txp.Core                 (Tx, TxId, topsortTxs, taTx, txpTxs)
+import           Pos.Txp.Toil.Class           (MonadUtxoRead)
+import           Pos.Txp.Toil.Trans           (evalToilTEmpty)
+import           Pos.Txp.Toil.Failure         (ToilVerFailure)
+import           Pos.Txp.Toil.Utxo.Functions  (VTxContext (..), verifyTxUtxo,
+                                               applyTxToUtxo)
+import           Pos.Wallet.Web.Pending.Types (PendingTx (..), PtxCondition (..),
+                                               ptxCond, _PtxApplying, _PtxWontApply,
+                                               ptxTxAux, ptxTxId)
+import           Pos.Util.Util                (getKeys)
+
+
+-- | Reevaluate all pending txs (in 'PtxApplying' and 'PtxWontApply' states):
+--
+--   * 'PtxInNewestBlocks' for ones that are already in the blockchain
+--     (at any depth)
+--   * 'PtxWontApply' for ones that can't be applied
+--   * 'PtxApplying' to ones that can be applied either directly or after
+--     applying some other transactions
+--
+-- TODO: do we need to do anything special to ensure that the UTXO we're
+-- shown doesn't include outputs from failed txs or mempool txs? I'm not
+-- sure
+reevaluateUncertainPtxs
+    :: forall ssc m.
+       (DB.MonadBlockDB ssc m, HasConfiguration, MonadUtxoRead m, WithLogger m)
+    => HashMap TxId PendingTx -> m (HashMap TxId PendingTx)
+reevaluateUncertainPtxs ptxs = do
```

Comment:
I am worried about this function, because DB can change while we are doing it. I'd like to see rocksdb snapshot or `StateLock` here. Perhaps `StateLock` is safer and easier, because we have never used rocksdb snapshots.